### ISSUE TITLE
How is the performance of classification task measured using BinaryClassificationEvaluator?

### ISSUE THREAD

**Issue by tonyrmit** [UKPLab/sentence-transformers#517]
Hi,
Thanks so much for the great project - SBERT! I am fine-tuning a classification task using the pre-trained `bert-base-uncased` in SBERT. I am trying to understand how the accuracy of binary classification (0 and 1) task is measured using `BinaryClassificationEvaluator`. When I looked at the code here:
https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/evaluation/BinaryClassificationEvaluator.py#L146
From my understanding, `acc` is calculating the percentage of positive labels (`ground-true` labels) as the `remaining_negatives` will be counted down to 0 eventually (at the end of the dev/test dataset). `best_threshold` would be the average of the cosine scores in the last two examples. I have a couple of questions here:
1,  I don't understand the way that is used to iterate the whole dataset to get the percentage of positive labels. Wouldn't we just do:
`max_acc = sum(labels == 1) / len(labels)`? And what the `max_acc` is used for and why? Sorry. I might mis-understood the code. It would be much appreciated if you could explain it a bit?
2, The `best_threshold` is the average cosine scores of last two examples? 
3, I would expect that `BinaryClassificationEvaluator` would calculate the accuracy between the predicted class (1 or 0) and ground-true labels for us so we can measure how well the trained model is doing. But looks like both `max_acc` and `best_threshold` are not what I want. Please correct me if I was wrong.

Thanks!

Regards,

Tony

**Comment by nreimers** [comment]
If your final task is classification, cross encoder might be the better option:
https://www.sbert.net/docs/usage/cross-encoder.html
https://www.sbert.net/examples/training/cross-encoder/README.html

The bi-encoder approach with SBERT is based on computing sentence emebddings independently for sentences. This can be helpful for example to identify duplicate questions in a large corpus: You compute the embeddings for all questions. Then you do the pair-wise cosine-similarity between all possible question pairs. Then you mark all questions with a cosine similarity higher than a threshold as duplicate.

This is kind of simulated in this evaluator: The cosine similarity between the embeddings is computed. However, as you do not know if you should mark all questions with a similarity above 0.5 as duplicate, or maybe with a score above 0.75 or maybe only pairs with similarity above 0.9 as "1" ,  it determines the optimal threshold.

So if you classify all pairs with similarity > best_threshold, then you will get the accuracy max_acc.

If you choose another threshold for the classification, the accuracy will be lower.

**Comment by tonyrmit** [comment]
Thanks @nreimers! I will try `cross-encoder` and see. 😄
